Sources/SwiftProtobuf/ProtobufBinaryTypes.swift

```diff
@@ -789,14 +777,21 @@ public extension ProtobufBinaryMessageBase {
 
 public extension ProtobufMessage {
     static func decodeProtobufMapValue(decoder: inout ProtobufFieldDecoder, value: inout Self?) throws {
-        let handled = try decoder.decodeSingularMessageField(fieldType: Self.self, value: &value)
-        assert(handled)
+        try decoder.decodeSingularMessageField(fieldType: Self.self, value: &value)
+        assert(value != nil)
     }
 
     init(protobuf: Data) throws {
         try self.init(protobuf: protobuf, extensions: nil)
     }
 
+    init(protobufBuffer: UnsafeBufferPointer<UInt8>) throws {
+        try self.init(protobufBuffer: protobufBuffer, extensions: nil)
+    }
+}
+
+// For Proto3 messages, we need to preserve the unknown fields when we're done
```

Comment:
(1) is definitely a concern. There might be a happy medium; keep `decodeIntoSelf` and then have it call something like `_protoc_gen_takeUnknowns` that gets generated to do nothing in proto3 (perhaps even a protocol extension default implementation) and gets generated to set the `unknown` property in proto3. It's still slightly more generated code, but less.

For (2), I'm not sure how much value there is in letting clients determine whether a message is proto2 or proto3. That would arguably be the role of the descriptor—if we were generating them—and the syntax is technically a property of the *file*, not of the *message*, even though it affects how the behavior of the message.

At least two of the language implementations (Obj-C and Java) don't bother distinguishing between proto2 and proto3 at all in the public message API (they just return null/empty extension sets and unknown stores for proto3, for example), so it's questionable whether we'd want to support something that doesn't appear to have a strong need. If there *were* cases where someone needed this information in a generic fashion, they can already do that using a visitor.